CustomDialogOnClose/explainer.md

```diff
@@ -0,0 +1,125 @@
+```javascript
+window.addEventListener("beforeunload", (event) => {
+    if (event.dialog) {
+      event.dialog.setMessage("You have unsaved changes, would you like to save them?");
+      event.dialog.setButtonLabel("Save");
+      event.dialog.show().then(async (result) => {
+        if (result == "Save"){
+          // save the document.
+        }
+      });
+    }
+});
+```
+
+![A dialog alerting the user that they have unsaved changes. Three options are presented "Save", "Leave", and "Cancel".](./images/CustomLeaveSiteExample.png)
+
+`show()` returns a promise that will resolve with the user's selection. Once it resolves, the code provided by the developer will run, and the tab will close when it is finished. While the developer's code is running, we should display a dialog indicating that some work is being done, and present an option to the user that let's them cancel the work and leave immediately. Here is an example of how this "Cleaning up" dialog may look:
+
+![A dialog with the title "Cleaning up" and the text "This page is cleaning up before it closes. You may close the page now, but you may experience data loss." There is also a button that says "Leave Now".](./images/CleaningUpExample.png)
+
+
+#### Time limit
+To avoid a poor user experience from websites that take a long time to complete their task, we propose that there should be a time limit on how long the operation can run for. Currently, Chromium based browser display a warning if the synchronous code in the beforeunload event handler takes a long time.
+
+`[Violation] 'beforeunload' handler took 53921ms`
+
+However, no time limit is actually enforced. We propose strictly enforcing a time limit that starts as soon as the user makes a selection on the "Leave Site?" dialog. What that limit should be will need to be researched further. An error message such as this should be displayed if the limit is reached:
+`The beforeunload promise did not resolve within the X second time limit. The operation has been cancelled.`
+
+#### Sequence Diagram
+Below is a sequence diagram to illustrate the flow when this new capability is used.
+```
+App         "Leave Site?" Dialog     "Cleaning up" Dialog     Developer's Promise
+ |
+ | User attempts to close app
+ | ----------------->| User selects "Save" option
+ | <---------------- |
+ | --------------------------------------> |
+ |                                         |-------------------> |                     
+ |                                         |                     | Unsaved changes are saved
+ |                                         | <------------------ |
+ | <-------------------------------------- | Promise resolves, times out, or user force exits
+ |  "Cleaning up" Dialog closes
+App closes
+```
+
+## Avoiding Misuse
```

Comment:
Stylistic, optional feedback: could be nice to display this as a two-column table for easy if => then scanning.